### PR TITLE
[python] Keep only the read or write TileDB handles open long-term

### PR DESCRIPTION
**Pull request by thetorpedodog**
Context: #638 and #540

This reorganizes the way TileDB handles are managed to ensure that they are only opened for as long as strictly necessary.

Both handle and metadata management are both moved into `handles.py`. The `Handle` object is renamed `Wrapper` (because otherwise it got very confusing) and its construction process has been revamped to store only the single raw TileDB handle it owns (along with metadata). For write handle (wrappers), any reading is done by a short-lived read handle opened onto the same TileDB object.

This also let me overhaul the way `factory.open` works, which I had way overcomplicated on the first go round. It is now much simpler.

---

Once again with apologies to @mlin. This used to reduce >100 lines of code but then I found a bug in metadata handling (that may have already existed?) and added a bunch of tests to characterize it. Then I found some duplicate code that could be merged and got it back to net-negative, which is something.

### CODE REVIEW

**Comment by thetorpedodog** [comment]
This change will also in part address https://github.com/single-cell-data/TileDB-SOMA/issues/539; with a couple test cases in a follow-up `factory.open` should be ready for export

**Comment by thetorpedodog** [comment]
I got a little lost in the sauce working on this so in the morning I’m going to spend a few minutes decomposing this into a few smaller commits as best I can. The overall diff should still be the same (maybe with a few more comments).

**Comment by thetorpedodog** [comment]
Finally, [I did it to myself](https://twitter.com/dril/status/873264183281889281):

This branch has conflicts that must be resolved

File: apis/python/src/tiledbsoma/tiledb_object.py [hunk 1]
```diff
@@ -23,9 +25,27 @@ class TileDBObject(somacore.SOMAObject, Generic[_HandleType]):
...
+    @classmethod
+    def open(
+        cls: Type[_Self],
+        uri: str,
+        mode: options.OpenMode = "r",
+        *,
+        context: Optional[SOMATileDBContext] = None,
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> _Self:
+        """Opens this specific type of collection."""
```
Comment (mlin) [inline apis/python/src/tiledbsoma/tiledb_object.py:37]: Is this docstring correct?

File: apis/python/src/tiledbsoma/handles.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
+"""Abstractions to more easily manage read and write access to TileDB data.
```
Comment (mlin) [inline apis/python/src/tiledbsoma/handles.py:1]: Filename- how about smth like `tiledb_handle_wrappers.py`?

**Review by mlin — APPROVED** [review]
LGTM subject to two nits inline

The nice clean factorization of the Wrapper.open logic is going to get a little messier when we add timestamps unfortunately, because of the way the timestamps have to be set in starkly different ways on group vs array. But I'm pretty sure we'll be able to contain that within the implementation details of ArrayWrapper/GroupWrapper, and hopefully TileDB-Py will make the group timestamps more straightforward in the future.

File: apis/python/src/tiledbsoma/tiledb_object.py (hunk 1 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/tiledb_object.py]: It is (but only when you are calling this on a Collection type).

File: apis/python/src/tiledbsoma/handles.py (hunk 2 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/handles.py]: I went with `tdb_handles.py` in the interest of keeping it short and reducing repetition.

**Pull request merged**